system/lib/wasmfs/backends/opfs_backend.cpp

```diff
@@ -0,0 +1,279 @@
+                       int file_id,
+                       int new_dir_id,
+                       const char* name);
+
+void _wasmfs_opfs_remove_child(em_proxying_ctx* ctx,
+                               int dir_id,
+                               const char* name);
+
+void _wasmfs_opfs_get_entries(em_proxying_ctx* ctx,
+                              int dirID,
+                              std::vector<Directory::Entry>* entries);
+
+void _wasmfs_opfs_open(em_proxying_ctx* ctx, int file_id, int* access_id);
+
+void _wasmfs_opfs_close(em_proxying_ctx* ctx, int access_id);
+
+void _wasmfs_opfs_free_file(int file_id);
+
+void _wasmfs_opfs_free_directory(int dir_id);
+
+// Synchronous read. Return the number of bytes read.
+int _wasmfs_opfs_read(int access_id, uint8_t* buf, uint32_t len, uint32_t pos);
+
+// Synchronous write. Return the number of bytes written.
+int _wasmfs_opfs_write(int access_id,
+                       const uint8_t* buf,
+                       uint32_t len,
+                       uint32_t pos);
+
+void _wasmfs_opfs_get_size(em_proxying_ctx* ctx, int access_id, uint32_t* size);
+
+void _wasmfs_opfs_set_size(em_proxying_ctx* ctx, int access_id, uint32_t size);
+
+void _wasmfs_opfs_flush(em_proxying_ctx* ctx, int access_id);
+
+} // extern "C"
+
+class OPFSFile : public DataFile {
+public:
+  ProxyWorker& proxy;
+
+  // The IDs of the corresponding file handle and, if the file is open, the
+  // corresponding access handle.
+  int fileID;
+  int accessID = -1;
+
+  // The number of times this file has been opened. We only close its
+  // AccessHandle when this falls to zero.
+  size_t openCount = 0;
+
+  OPFSFile(mode_t mode, backend_t backend, int fileID, ProxyWorker& proxy)
+    : DataFile(mode, backend), fileID(fileID), proxy(proxy) {}
+
+  ~OPFSFile() override {
+    assert(openCount == 0);
+    assert(accessID == -1);
+    proxy([&]() { _wasmfs_opfs_free_file(fileID); });
+  }
+
+private:
+  size_t getSize() override {
+    uint32_t size;
+    proxy([&](auto ctx) { _wasmfs_opfs_get_size(ctx.ctx, accessID, &size); });
+    return size_t(size);
+  }
+
+  void setSize(size_t size) override {
+    proxy([&](auto ctx) { _wasmfs_opfs_set_size(ctx.ctx, accessID, size); });
+  }
+
+  void open(oflags_t flags) override {
+    if (openCount == 0) {
+      proxy([&](auto ctx) { _wasmfs_opfs_open(ctx.ctx, fileID, &accessID); });
+      ++openCount;
+    }
+    // TODO: proper error handling.
+    assert(accessID >= 0);
+  }
+
+  void close() override {
+    assert(openCount >= 1);
+    if (--openCount == 0) {
+      proxy([&](auto ctx) { _wasmfs_opfs_close(ctx.ctx, accessID); });
+      accessID = -1;
+    }
+  }
+
+  __wasi_errno_t read(uint8_t* buf, size_t len, off_t offset) override {
+    uint32_t nread;
+    proxy([&]() { nread = _wasmfs_opfs_read(accessID, buf, len, offset); });
+    // TODO: Add a way to report the actual bytes read. We currently assume the
+    // available bytes can't change under us.
```

Comment:
```suggestion
    // available bytes can't change under us in OPFS.
```
Less of a suggestion and more checking that I understand the concern here - is that right?

Reply:
No, this is a problem with the WasmFS syscall implementation. It currently reads the size of the file to determine how many bytes should be available to read before issuing `read` calls, which is both unnecessary work and incorrect in the face of concurrent modifications. This is only a performance problem for the OPFS backend since concurrent modifications to the file are impossible, but it is also arguably a correctness problem for the node backend where we do want to allow some limited form of concurrent file system modification.